collectors/ebpf.plugin/ebpf_apps.c

```diff
@@ -454,3 +454,588 @@ int ebpf_read_apps_groups_conf(struct target **apps_groups_default_target, struc
+        if(unlikely(!p->sortlist))
+            p->sortlist = sortlist++;
+    }
+
+    if(all_pids[1])
+        all_pids[1]->sortlist = sortlist++;
+
+    // give a target to all merged child processes
+    found = 1;
+    while(found) {
+        if(unlikely(debug_enabled)) loops++;
+        found = 0;
+        for(p = root_of_pids; p ; p = p->next) {
+            if(unlikely(!p->target && p->merged && p->parent && p->parent->target)) {
+                p->target = p->parent->target;
+                found++;
+
+                if(debug_enabled || (p->target && p->target->debug_enabled))
+                    debug_log_int("TARGET INHERITANCE: %s is inherited by %d (%s) from its parent %d (%s) at phase 2.", p->target->name, p->pid, p->comm, p->parent->pid, p->parent->comm);
+            }
+        }
+    }
+
+    debug_log("apply_apps_groups_targets_inheritance() made %d loops on the process tree", loops);
+}
+
+static inline void post_aggregate_targets(struct target *root) {
+    struct target *w;
+    for (w = root; w ; w = w->next) {
+        if(w->collected_starttime) {
+            if (!w->starttime || w->collected_starttime < w->starttime) {
+                w->starttime = w->collected_starttime;
+            }
+        } else {
+            w->starttime = 0;
+        }
+    }
+}
+
+static inline void del_pid_entry(pid_t pid) {
+    struct pid_stat *p = all_pids[pid];
+
+    if(unlikely(!p)) {
+        error("attempted to free pid %d that is not allocated.", pid);
+        return;
+    }
+
+    debug_log("process %d %s exited, deleting it.", pid, p->comm);
+
+    if(root_of_pids == p)
+        root_of_pids = p->next;
+
+    if(p->next) p->next->prev = p->prev;
+    if(p->prev) p->prev->next = p->next;
+
+    freez(p->stat_filename);
+    freez(p->status_filename);
+    freez(p->io_filename);
+    freez(p->cmdline_filename);
+    freez(p->cmdline);
+    freez(p);
+
+    all_pids[pid] = NULL;
+    all_pids_count--;
+}
+
+static void cleanup_exited_pids(void) {
+    struct pid_stat *p = NULL;
+
+    for(p = root_of_pids; p ;) {
+        if(!p->updated && (!p->keep || p->keeploops > 0)) {
+            if(unlikely(debug_enabled && (p->keep || p->keeploops)))
+                debug_log(" > CLEANUP cannot keep exited process %d (%s) anymore - removing it.", p->pid, p->comm);
+
+            pid_t r = p->pid;
+            p = p->next;
+            del_pid_entry(r);
+        }
+        else {
+            if(unlikely(p->keep)) p->keeploops++;
+            p->keep = 0;
+            p = p->next;
+        }
+    }
+}
+
+int collect_data_for_all_processes(
+    int (*bpf_map_get_next_key)(int, const void *, void *), int (*bpf_map_lookup_elem)(int, const void *, void *),
+    int tbl_pid_stats_fd)
+{
+    pid_t key, next_key;
+    ebpf_process_stat_t *value = NULL;
+
+    while (bpf_map_get_next_key(tbl_pid_stats_fd, &key, &next_key) == 0) {
+        if (!bpf_map_lookup_elem(tbl_pid_stats_fd, &next_key, value)) {
+            if (value)
```

Comment:
I do not think this is necessary, because we only have a key stored on any eBPF hash table when a data is stored, we have a kind of 1 <- > 1 relationship between them.